File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, INT_MAX);
        dp[0]=0;
        // sort(coins.begin(), coins.end());
        for(int i=1; i<= amount; i++){
            int amt = i;
            for(auto coin : coins){
                if(coin> amt)continue;
                // if(dp[i-1]+ coin == amt){
                //     int ls = dp[i-1]+1;
                // }
                if(dp[amt-coin]!= INT_MAX)
                dp[i]= min(dp[amt-coin]+1, dp[i]);
            }
        }
        return dp[amount]!=INT_MAX?dp[amount]:-1;
    }
};

File: 0322-coin-change/0322-coin-change.rs
impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let max = i32::MAX;
        let mut dp = vec![max; (amount + 1) as usize];
        dp[0] = 0;

        for i in 1..=amount as usize {
            for &coin in &coins { // change here, borrowing 'coins' for iteration
                if coin as usize <= i && dp[i - coin as usize] != max {
                    dp[i] = dp[i].min(dp[i - coin as usize] + 1);
                }
            }
        }

        if dp[amount as usize] == max { 
            -1 
        } else { 
            dp[amount as usize] 
        }
    }
}